File: client/startup_dialog.py
import socket

from PySide6.QtWidgets import QDialog, QVBoxLayout, QHBoxLayout, QLabel, QLineEdit, QPushButton, QMessageBox
from PySide6.QtCore import Qt

CONTROL_PORT = 50001
DSCP_CS3 = 24
IP_TOS_CS3 = DSCP_CS3 << 2


class ServerIPDialog(QDialog):
    """Dialog for manual server IP entry"""
    
    def __init__(self):
        super().__init__()
        self.server_ip = None
        
        self.setWindowTitle("Voice App - Server Not Found")
        self.setGeometry(100, 100, 400, 150)
        
        layout = QVBoxLayout()
        
        layout.addWidget(QLabel("Server not found via broadcast."))
        layout.addWidget(QLabel("Enter server IP address:"))
        
        self.ip_input = QLineEdit()
        self.ip_input.setPlaceholderText("192.168.0.X")
        layout.addWidget(self.ip_input)
        
        btn_layout = QHBoxLayout()
        
        ok_btn = QPushButton("Connect")
        ok_btn.clicked.connect(self.accept)
        btn_layout.addWidget(ok_btn)
        
        cancel_btn = QPushButton("Cancel")
        cancel_btn.clicked.connect(self.reject)
        btn_layout.addWidget(cancel_btn)
        
        layout.addLayout(btn_layout)
        self.setLayout(layout)
    
    def accept(self):
        ip = self.ip_input.text().strip()
        if ip:
            self.server_ip = ip
            super().accept()


class StartupDialog(QDialog):
    """Dialog to enter unique client name and confirm server"""
    
    def __init__(self, server_ip, audio_port):
        super().__init__()
        self.server_ip = server_ip
        self.audio_port = audio_port
        self.client_id = None
        
        self.setWindowTitle("Voice App - Client Setup")
        self.setGeometry(100, 100, 400, 200)
        
        layout = QVBoxLayout()
        
        # Server IP display
        layout.addWidget(QLabel(f"Server: {server_ip}"))
        layout.addWidget(QLabel(f"Audio Port: {audio_port}"))
        
        # Unique name input
        layout.addWidget(QLabel("Enter your unique name:"))
        self.name_input = QLineEdit()
        self.name_input.setPlaceholderText("e.g. Alice / ControlRoom-1")
        self.name_input.returnPressed.connect(self.accept)
        layout.addWidget(self.name_input)
        
        # Buttons
        btn_layout = QHBoxLayout()
        
        ok_btn = QPushButton("OK")
        ok_btn.clicked.connect(self.accept)
        btn_layout.addWidget(ok_btn)
        
        cancel_btn = QPushButton("Cancel")
        cancel_btn.clicked.connect(self.reject)
        btn_layout.addWidget(cancel_btn)
        
        layout.addLayout(btn_layout)
        self.setLayout(layout)
    
    def accept(self):
        candidate = self.name_input.text().strip()

        if not candidate:
            msg = QMessageBox(self)
            msg.setIcon(QMessageBox.Warning)
            msg.setWindowTitle("Invalid Name")
            msg.setText("Name cannot be empty.")
            msg.exec()
            return

        if ":" in candidate:
            msg = QMessageBox(self)
            msg.setIcon(QMessageBox.Warning)
            msg.setWindowTitle("Invalid Name")
            msg.setText("Name cannot contain ':'.")
            msg.exec()
            return

        if not self._is_name_available(candidate):
            msg = QMessageBox(self)
            msg.setIcon(QMessageBox.Warning)
            msg.setWindowTitle("Name Already In Use")
            msg.setText(f"'{candidate}' is already connected.")
            msg.setInformativeText("Please choose a different name.")
            msg.exec()
            return

        self.client_id = candidate
        super().accept()

    def _is_name_available(self, candidate):
        ctrl = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            ctrl.setsockopt(socket.IPPROTO_IP, socket.IP_TOS, IP_TOS_CS3)
            ctrl.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
        except OSError:
            pass
        ctrl.settimeout(2.5)
        try:
            ctrl.connect((self.server_ip, CONTROL_PORT))
            ctrl.sendall(b"LIST\n")
            response = ctrl.recv(1024).decode(errors="ignore").strip()
            if not response:
                return True
            existing = {name.strip() for name in response.split(",") if name.strip()}
            return candidate not in existing
        except Exception:
            # If list check fails, registration flow in main.py still enforces uniqueness.
            return True
        finally:
            ctrl.close()

File: client/startup_dialog.cpp
#include "startup_dialog.h"

#include "control_client.h"

#include <QHBoxLayout>
#include <QIcon>
#include <QLabel>
#include <QLineEdit>
#include <QMessageBox>
#include <QPushButton>
#include <QVBoxLayout>

// No additional networking needed here; control checks handled via send_control_command.

ServerIPDialog::ServerIPDialog(QWidget* parent)
    : QDialog(parent) {
    setWindowTitle("Voice App - Server Not Found");
    setFixedSize(400, 150);

    auto* layout = new QVBoxLayout(this);
    layout->addWidget(new QLabel("Server not found via broadcast."));
    layout->addWidget(new QLabel("Enter server IP address:"));

    ip_input_ = new QLineEdit(this);
    ip_input_->setPlaceholderText("192.168.0.X");
    layout->addWidget(ip_input_);

    auto* btn_layout = new QHBoxLayout();
    auto* ok_btn = new QPushButton("Connect", this);
    auto* cancel_btn = new QPushButton("Cancel", this);
    connect(ok_btn, &QPushButton::clicked, this, &ServerIPDialog::accept);
    connect(cancel_btn, &QPushButton::clicked, this, &ServerIPDialog::reject);
    btn_layout->addWidget(ok_btn);
    btn_layout->addWidget(cancel_btn);
    layout->addLayout(btn_layout);
}

void ServerIPDialog::accept() {
    const QString ip = ip_input_->text().trimmed();
    if (!ip.isEmpty()) {
        server_ip_ = ip;
        QDialog::accept();
    }
}

StartupDialog::StartupDialog(const QString& server_ip, int audio_port, QWidget* parent)
    : QDialog(parent), server_ip_(server_ip), audio_port_(audio_port) {
    setWindowTitle("Voice App - Client Setup");
    setFixedSize(400, 200);

    auto* layout = new QVBoxLayout(this);
    layout->addWidget(new QLabel(QString("Server: %1").arg(server_ip_)));
    layout->addWidget(new QLabel(QString("Audio Port: %1").arg(audio_port_)));
    layout->addWidget(new QLabel("Enter your unique name:"));

    name_input_ = new QLineEdit(this);
    name_input_->setPlaceholderText("e.g. Alice / ControlRoom-1");
    connect(name_input_, &QLineEdit::returnPressed, this, &StartupDialog::accept);
    layout->addWidget(name_input_);

    auto* btn_layout = new QHBoxLayout();
    auto* ok_btn = new QPushButton("OK", this);
    auto* cancel_btn = new QPushButton("Cancel", this);
    connect(ok_btn, &QPushButton::clicked, this, &StartupDialog::accept);
    connect(cancel_btn, &QPushButton::clicked, this, &StartupDialog::reject);
    btn_layout->addWidget(ok_btn);
    btn_layout->addWidget(cancel_btn);
    layout->addLayout(btn_layout);
}

void StartupDialog::accept() {
    QString candidate = name_input_->text().trimmed();
    candidate = candidate.endsWith(',') ? candidate.left(candidate.size() - 1) : candidate;

    if (candidate.isEmpty()) {
        QMessageBox::warning(this, "Invalid Name", "Name cannot be empty.");
        return;
    }

    if (candidate.contains(':') || candidate.contains('|') || candidate.contains(',') || candidate.contains('\n') || candidate.contains('\r') || candidate.contains('\t')) {
        QMessageBox::warning(this, "Invalid Name", "Name cannot contain ':', '|', ',' or control characters.");
        return;
    }

    if (!isNameAvailable(candidate)) {
        QMessageBox::warning(this, "Name Already In Use", QString("'%1' is already connected.\nPlease choose a different name.").arg(candidate));
        return;
    }

    client_id_ = candidate;
    QDialog::accept();
}

bool StartupDialog::isNameAvailable(const QString& candidate) {
    const auto resp = send_control_command(server_ip_.toStdString(), "LIST", 2500);
    if (!resp.ok) {
        return true;
    }
    const QStringList existing = parseListResponse(QString::fromStdString(resp.response));
    return !existing.contains(candidate);
}

QStringList StartupDialog::parseListResponse(const QString& response) {
    if (response.isEmpty()) {
        return {};
    }
    if (response.contains('\n')) {
        return response.split('\n', Qt::SkipEmptyParts);
    }
    return response.split(',', Qt::SkipEmptyParts);
}
